enhancements/authentication/custom-scc-preemption-prevention.md

```diff
@@ -0,0 +1,260 @@
+
+## Summary
+
+OpenShift ships SCCs by default that applies both to core OpenShift workloads
+as well as to user provided workloads.
+
+Users have the possibility to add custom SCCs to OpenShift.
+These can impact core OpenShift workloads and, worst case, break out of the box components.
+
+This enhancement proposal suggests a mechanism to be able to protect core OpenShift
+out of the box workloads from side effects induced by custom SCCs.
+
+## Motivation
+
+In many cases the default out of the box SCCs are sufficient as they cover the most
+common patterns for controlling permissions pods can have.
+Additionally, the most recent enablement of Pod Security Admission also covers
+an additional layer of protection next to OpenShift SCCs.
+
+However, OpensShift historically does allow cluster administrators to add
+additional custom SCCs on top of the default out of the box SCCs.
+Custom SCCs are not considered different from the perspective of
+the SCC admission plugin.
+
+When a pod is being admitted, the admission plugin lists all SCCs available in the system, including all custom SCCs.
+The list of SCCs is then sorted by:
+- SCC priority in descending order
+- SCCs with the same priority: SCC restriction score in descending order (most restrictive first, least restrictive last)
+- SCCs with the same restriction score: lexical ordering of the SCC name in descending order
+
+An SCC is considered applicable if the requesting user or the
+pod's service account has `use` permissions against the `security.openshift.io.securitycontextconstraints` resource
+for an SCC in the requested namespace.
+If an SCC is applicable, the pod's manifest is potentially mutated according to the SCC logic.
+
+One can see that custom SCCs can easily overrule existing out of the box SCCs
+if the pod's service account or the requesting user has sufficient privileges to use any of them.
+
+In that case, if a custom SCC has higher priority, is more restrictive, or it's name is lexically ordered before an out of the box SCC,
+any OpenShift out of the box core pod will be mutated according to the custom SCC specified behavior.
+
+This can lead to unknown mutations of core OpenShift workloads.
+Worst case it can render OpenShift core workloads dysfunctional
+if i.e. a custom SCC is more restrictive than a previously applicable less restrictive out of the box SCC that was applicable to that workload.
+
+This proposal includes two suggested protection modes, a preferred "hard protection" and a "soft protection" variant
+which are described below.
+
+We need a mechanism to protect OpenShift core workloads such that they are not
+subject to pod mutation due to user provided custom SCCs.
+
+### User Stories
+
+1. As an OpenShift core developer I want to make sure that my shipped workload
+is guaranteed to be associated with an out of the box SCC.
+
+2. As an OpenShift core developer I want to make sure that my shipped workload
+is not preempted by custom user provided SCCs.
+
```

Comment:
since the goals include making this usable for end users, the use cases shouldn't be restricted to "openshift core developers" and "shipped workloads"

Reply:
oh right, fixed